pkg/redirectpolicy/manager.go

```diff
@@ -672,22 +661,53 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
 				}
 			}
 			if len(bes4) > 0 {
-				rpm.upsertServiceWithBackends(config, feM, pod.id, bes4)
+				rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
 			} else if len(bes6) > 0 {
-				rpm.upsertServiceWithBackends(config, feM, pod.id, bes6)
+				rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
 			}
 		}
 	}
 }
 
-// upsertServiceWithBackends updates policy config internal state and upserts
-// service with the given pod backends.
-func (rpm *Manager) upsertServiceWithBackends(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
-	frontendMapping.backends = backends
-	rpm.policyPods[podID] = append(rpm.policyPods[podID], podPolicyInfo{
-		policyID: config.id,
-		backends: backends,
-	})
+// upsertFrontendMapping updates policy config internal state and upserts
+// service with the policy frontend and mapped backends.
+func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
+	newFePods := make([]backend, 0, len(frontendMapping.podBackends))
+	// Update the frontend mapped backends slice, keeping the order same.
+	// TODO (aditi) unit test case
+	updatePodBes := true
+	for _, be := range frontendMapping.podBackends {
+		if be.podID == podID {
+			if updatePodBes {
+				updatePodBes = false
+				// Get the updated backends for the given pod.
+				newFePods = append(newFePods, backends...)
+			}
+		} else {
+			// Collect the unchanged backends for other pods.
+			newFePods = append(newFePods, be)
```

Comment:
I'm a bit confused why the caller doesn't just pass down the frontend and all desired backends, why do we need to fold in the previous `frontendMapping.podBackends` here? The function is called `upsertFrontendMapping()`, ie "update or insert this set", not "merge" or "append".

Reply:
We don't compute delta when pod backends are updated. We'll override with the current service entry for a frontend with the updated frontend's backends list.

Reply:
We don't compute delta for pod backend updates. Whenever frontend's pod backends change, we will override the existing service entry with the updated backends. The "upsert" name looks fine to me, if there are currently no backends for a frontend, then it'll insert the new pod backends, else it'll update the backends slice with the pod backends. Also, keep in mind that we need to maintain the frontend to backends mapping (e.g., fe tcp-> be tcp, fe udp->be udp).

Reply:
My thought on this was not fully developed when I previously asked the question, so bear with me. Here's my understanding of the overall logic going on here:

* `processConfigWithSinglePort()` / `processConfigWithNamedPorts()` loop through the pods
  * For each pod, it collects the backends and calls `upsertFrontendMapping()`
    * This logic folds the new backends into the current `frontendMapping` then calls `rpm.UpsertService()` to call down to the datapath

Why not have the following:
* `processConfigWithSinglePort()` / `processConfigWithNamedPorts()` loop through the pods to collect the backends
* Then invoke `upsertFrontendMapping()` once (through to `rpm.UpsertService()` once) with all backends

Would this allow us to simplify the logic in this function?

Reply:
>Also, keep in mind that we need to maintain the frontend to backends mapping (e.g., fe tcp-> be tcp, fe udp->be udp).

We need to call `rpm.upsertFrontendMapping` and `rpm.upsertService` for every frontend. If you feel we are still not on the same page, I'm happy to discuss it offline in more details.

Reply:
I agree that we need to call both of these for every frontend, my comment above is referring to backends (the inner loop in `processConfigWithNamedPorts()`), not the frontends (the outer loop in `processConfigWithNamedPorts()`).

Reply:
processConfigWith* processes updates for a given pod, and then we aggregate all the backend pods for a frontend with `upsertFrontendMapping`. I'm not sure which part can be simplified here.

Reply:
I guess I just don't understand why we can't aggregate the backend pods in processConfigWith* into a simple slice and call `upsertFrontendMapping()` once with all backends at once.

As far as I understand at the moment the code does:
* Iterate through all frontends (processConfigWith*)
  * Iterate through all backends (processConfigWith*)
    * Iterate through all previously-iterated backends (`upsertFrontendMapping()`)

If processConfigWith* collected all backends and invoked `upsertFrontendMapping()` once per frontend, then the third loop could be removed, and furthermore we would avoid issuing a `upsertService()` once per backend and instead issue once per frontend.

Reply:
The data structures that you suggested might not be amenable for that sort of a change.
Currently, `upsertFrontendMapping` receives only new pod backends, which are then merged with the other backends of the frontend. If we aggregate all the pod backend updates in a slice, we don't know which pods have been updated, and we would still need a 2nd inner loop.

Iterate through all frontends (processConfigWith*)
   Iterate through all backends (processConfigWith*)
       Collect pod backend updates in a slice
   Iterate through all frontend backends
      Iterate through the pod backend updates slice


I'll need to introduce some other data structures and maintain additional state in there. 
This could work - 

```
type podBeUpdate {
    bes []backend
    updatePodBe bool
}
```
podUpdatesMap := make(map[podID][]podBeUpdate)
iterate through all pod backends
    Add backends to the map

Iterate through all frontend backends
    if be.podID is in the map
       Override the backends for this podID (flip the updatePodBe flag to false so that subsequent entries can be skipped)
    else
       Add the unchanged pod be

Override frontend backends


This is mostly what we are doing in `upsertFrontendMapping`, but with more simplicity. IMO, with the current scale, I don't see much value in adding the additional complexity at this time.

Reply:
As a middle ground, I can move `upsertService` out of `upsertFrontendMapping` so that it's called just once for every frontend.

Reply:
Done. PTAL.

Reply:
> As a middle ground, I can move upsertService out of upsertFrontendMapping so that it's called just once for every frontend.

That's what I was trying (but apparently failing) to communicate :sweat_smile: 

When I started this thread, I thought that the backend management (upper portion) in this function + the upsertService only made sense to run once per frontend, but I didn't really understand what the `rpm.policyPods` management portion did. Having dug a little deeper, I can see that the policyPods mapping update portion needs to be invoked for every pod that is touched by the policy. That seems to be the piece that I was failing to understand above. FWIW I toyed with the code and here's the incremental diff that achieves what I had in mind:

```
diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index e83bfd19f87d..e1fbbd1929c8 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -75,7 +75,7 @@ type Manager struct {
        // Stores mapping of redirect policy serviceID to the corresponding policyID for
        // easy lookup in policyConfigs
        policyServices map[k8s.ServiceID]policyID
-       // Stores mapping of pods to redirect policies that select the pods
+       // Stores desired mapping of pods to redirect policies that select the pods
        policyPods map[podID][]policyID
        // Stores redirect policy configs indexed by policyID
        policyConfigs map[policyID]*LRPConfig
@@ -615,13 +615,14 @@ func (rpm *Manager) processConfigWithSinglePort(config *LRPConfig, pods ...*podM
                                        bes6 = append(bes6, be)
                                }
                        }
-               }
-               if len(bes4) > 0 {
-                       rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
-               } else if len(bes6) > 0 {
-                       rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
+                       rpm.updatePodPolicyMapping(config.id, pod.id)
                }
        }
+       if len(bes4) > 0 {
+               rpm.upsertFrontendMapping(config, feM, bes4)
+       } else if len(bes6) > 0 {
+               rpm.upsertFrontendMapping(config, feM, bes6)
+       }
        return
 }

@@ -675,59 +676,43 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
                                                        bes6 = append(bes6, be)
                                                }
                                        }
+                                       rpm.updatePodPolicyMapping(config.id, pod.id)
                                }
                        }
-                       if len(bes4) > 0 {
-                               rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
-                       } else if len(bes6) > 0 {
-                               rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
-                       }
                }
-       }
-}
-
-// upsertFrontendMapping updates policy config internal state and upserts
-// service with the policy frontend and mapped backends.
-func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
-       newFePods := make([]backend, 0, len(frontendMapping.podBackends)+len(backends))
-       updatePodBes := true
-       // Update the frontend mapped backends slice, keeping the order same.
-       for _, be := range frontendMapping.podBackends {
-               if be.podID == podID {
-                       if updatePodBes {
-                               updatePodBes = false
-                               // Get the updated backends for the given pod.
-                               newFePods = append(newFePods, backends...)
-                       }
-               } else {
-                       // Collect the unchanged backends for other pods.
-                       newFePods = append(newFePods, be)
+               if len(bes4) > 0 {
+                       rpm.upsertFrontendMapping(config, feM, bes4)
+               } else if len(bes6) > 0 {
+                       rpm.upsertFrontendMapping(config, feM, bes6)
                }
        }
-       if updatePodBes {
-               // New backend pod for the frontend
-               newFePods = append(newFePods, backends...)
-       }
-       frontendMapping.podBackends = newFePods
+}

+func (rpm *Manager) updatePodPolicyMapping(policy policyID, podID podID) {
        if podPolicies, ok := rpm.policyPods[podID]; ok {
                newPodPolicy := true
                for _, poID := range podPolicies {
                        // Existing pod policy update
-                       if poID == config.id {
+                       if poID == policy {
                                newPodPolicy = false
                                break
                        }
                }
                if newPodPolicy {
                        // Pod selected by a new policy
-                       rpm.policyPods[podID] = append(rpm.policyPods[podID], config.id)
+                       rpm.policyPods[podID] = append(rpm.policyPods[podID], policy)
                }
        } else {
                // Pod selected by a policy for the first time
-               pp := []policyID{config.id}
+               pp := []policyID{policy}
                rpm.policyPods[podID] = pp
        }
+}
+
+// upsertFrontendMapping updates policy config internal state and upserts
+// service with the policy frontend and mapped backends.
+func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, backends []backend) {
+       frontendMapping.podBackends = backends
        rpm.upsertService(config, frontendMapping)

 ```

It's a little bit simpler (no more backends merging logic) and it now only issues an `UpsertService()` once per frontend. Though when I try it out, I notice some unit test failures, looks like the backends are being added to `frontendMapping` too many times.

Reply:
>It's a little bit simpler (no more backends merging logic) and it now only issues an UpsertService() once per frontend. Though when I try it out, I notice some unit test failures, looks like the backends are being added to frontendMapping too many times.

Yay, looks like the unit tests did their job. :) Anyway the latest commit that I addressed this, and the `upsertService` is being called only once per frontend.